### PR TITLE
Fix IndexMap

### PR DESCRIPTION
**Pull request by rozza**
Fixed issue converting from Range to Hash based Index maps
Fixed Range based lookup for last key

JAVA-1885

[Review on Reviewable](https://reviewable.io/reviews/rozza/mongo-java-driver/97)

### CODE REVIEW

**Comment by jyemin** [comment]
Initial review complete.

**Comment by rozza** [comment]
Updated based on feedback.

File: driver-core/src/test/unit/com/mongodb/internal/connection/IndexMapSpecification.groovy [hunk 1]
```diff
@@ -81,4 +81,33 @@ class IndexMapSpecification extends Specification {
...
+
+    def 'should include ranges when converting from range based to hash based indexMap'() {
+        given:
+        def indexMap = IndexMap.create(1000, 3)
+
+        when: 'converts from range based with a high startIndex to hash based'
+        indexMap = indexMap.add(5, 1005)
+
+        then:
+        1000 == indexMap.map(0)
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/internal/connection/IndexMapSpecification.groovy]: To test that loop condition better, let's assert that indexMap.map(-1) and indexMap.map(3) both throw MongoInternalException

**Comment by rozza** [comment]
Fixed codenarc, ported sync test to async and merged the mappings throws error tests.

**Pull request closed without merging**